src/mlpack/methods/ann/layer/convolution_impl.hpp

```diff
@@ -314,62 +306,90 @@ void Convolution<
 {
   arma::cube mappedError(((arma::Mat<eT>&) error).memptr(), outputWidth,
       outputHeight, outSize * batchSize, false, false);
-  arma::cube inputTemp(((arma::Mat<eT>&) input).memptr(), inputWidth,
-      inputHeight, inSize * batchSize, false, false);
+  arma::cube inputTemp;
+  if (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0) {
+    inputTemp = inputPaddedTemp;
+  } else {
+    inputTemp = arma::cube(((arma::Mat<eT>&) input).memptr(), inputWidth,
+                           inputHeight, inSize * batchSize, false, false);
+  }
 
   gradient.set_size(weights.n_elem, 1);
   gradientTemp = arma::Cube<eT>(gradient.memptr(), weight.n_rows,
       weight.n_cols, weight.n_slices, false, false);
   gradientTemp.zeros();
-
-  for (size_t outMap = 0, outMapIdx = 0, batchCount = 0; outMap <
-      outSize * batchSize; outMap++)
+  #pragma omp parallel for
+  for (omp_size_t outMapIdx = 0; outMapIdx < outSize; outMapIdx++)
   {
-    if (outMap != 0 && outMap % outSize == 0)
+    for (size_t inMap = 0; inMap < inSize; inMap++)
     {
-      batchCount++;
-      outMapIdx = 0;
+      arma::Mat<eT> &curGradTemp = gradientTemp.slice(outMapIdx+inMap);
```

Comment:
Just a tiny style issue. There are the same issues below.
```suggestion
      arma::Mat<eT>& curGradTemp = gradientTemp.slice(outMapIdx+inMap);
```